features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdh_alt.c

```diff
@@ -0,0 +1,256 @@
+          pHeap = ( uint8_t* )kgParams;
+          heapSize = sizeof(cc_ecc_25519_keygen_params_t);
+
+          ret = convert_CrysError_to_mbedtls_err( CRYS_ECMONT_KeyPair( kgParams->pubKey, ( size_t* )&public_key_size, kgParams->privKey,
+                                    &priv_key_size, &cc_rand, convert_mbedtls_to_cc_rand,
+                                    &kgParams->kgTempData ) );
+          if( ret != 0 )
+          {
+              goto cleanup;
+          }
+
+          MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( d, kgParams->privKey, priv_key_size ) );
+          MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary(  &Q->X, kgParams->pubKey, public_key_size ) );
+          MBEDTLS_MPI_CHK( mbedtls_mpi_lset( &Q->Z, 1 ) );
+      }
+      else
+          ret =  MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+
+cleanup:
+
+    if ( pHeap )
+    {
+        mbedtls_zeroize( pHeap, heapSize );
+        mbedtls_free( pHeap );
+    }
+
+    return ( ret );
+}
+#endif /* MBEDTLS_ECDH_GEN_PUBLIC_ALT */
+
+/*
+ * Compute shared secret (SEC1 3.3.1)
+ */
+#if defined (MBEDTLS_ECDH_COMPUTE_SHARED_ALT)
+int mbedtls_ecdh_compute_shared( mbedtls_ecp_group *grp, mbedtls_mpi *z,
+                         const mbedtls_ecp_point *Q, const mbedtls_mpi *d,
+                         int (*f_rng)(void *, unsigned char *, size_t),
+                         void *p_rng )
+{
+    int ret;
+    void* pHeap = NULL;
+    size_t   heapSize = 0;
+
+    size_t   public_key_size = (grp->nbits+7)/8 ;
+    const CRYS_ECPKI_Domain_t*  pDomain =  CRYS_ECPKI_GetEcDomain ( convert_mbedtls_grp_id_to_crys_domain_id( grp->id ) );
+    uint32_t secret_size =  ( ( grp->nbits + 7 ) / 8 ) ;
+    const uint32_t secret_size_in_heap = secret_size;
+    uint8_t* secret = mbedtls_calloc( 1, secret_size_in_heap );
+    if ( secret == NULL )
+        return ( MBEDTLS_ERR_ECP_ALLOC_FAILED );
+
+    /*
+     * Make sure Q is a valid pubkey before using it
+     */
+    MBEDTLS_MPI_CHK( mbedtls_ecp_check_pubkey( grp, Q ) );
+    if ( pDomain )
+    {
+        uint8_t  temp_buf[ 2 * MAX_KEY_SIZE_IN_BYTES + 1 ] = {0};
+        cc_ecc_ws_comp_shared_params_t* ecdhParams =  mbedtls_calloc( 1, sizeof(cc_ecc_ws_comp_shared_params_t) );
+
+        if ( ecdhParams == NULL )
+        {
+            ret = MBEDTLS_ERR_ECP_ALLOC_FAILED;
+            goto cleanup;
+        }
+
+        pHeap = ecdhParams;
+        heapSize = sizeof(cc_ecc_ws_comp_shared_params_t);
+
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_point_write_binary( grp, Q, MBEDTLS_ECP_PF_UNCOMPRESSED,
+                                                &public_key_size, temp_buf, sizeof(temp_buf) ) );
+
+        ret = convert_CrysError_to_mbedtls_err( CRYS_ECPKI_BuildPublKey( pDomain, temp_buf, public_key_size,
+                                                                         &ecdhParams->pubKey ) );
+        if ( ret != 0 )
+        {
+            goto cleanup;
+        }
+
+        memset ( temp_buf, 0, sizeof(temp_buf)  );
```

Comment:
Is it necessary to clear the public key information?

Reply:
yes, it is meant to avoid any unneeded information in the buffer, before writing private key information to it. See my answer to you in https://github.com/ARMmbed/mbed-os-cc-porting/pull/7#discussion_r139946103